File: Android/app/src/main/java/com/alafia/android/views/main/ClinicianMode.kt
package com.alafia.android.views.main

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

/**
 * Roles that can practise in clinician mode.
 *
 * This list lives here rather than inside a screen because three places need
 * it — the tab shell, the More grid and the Role screen — and a copy in each is
 * how they drift apart.
 */
val CLINICIAN_ROLES = setOf(
    "physician", "surgeon", "nurse_practitioner",
    "physician_assistant", "resident", "fellow", "attending_physician",
    "cardiologist", "dermatologist", "endocrinologist", "gastroenterologist",
    "neurologist", "oncologist", "pediatrician", "radiologist",
    "general_surgeon", "orthopedic_surgeon", "neurosurgeon",
    "cardiothoracic_surgeon", "plastic_surgeon", "vascular_surgeon",
    "oral_surgeon", "clinical_nurse_specialist", "nurse_anesthetist",
    "nurse_midwife", "charge_nurse", "nurse_administrator",
    "medical_director", "chief_medical_officer",
)

/**
 * Which persona the app is presenting: the user's own record, or their clinical
 * practice. Switching swaps the whole tab bar rather than adding a screen, so a
 * physician reviewing patients is not navigating past their own meal diary.
 *
 * Process-scoped rather than persisted: the mode lasts for the session and
 * resets on a cold start, which also means a signed-out clinician cannot leave
 * the next account in a mode its roles do not allow.
 */
object ClinicianModeState {
    var isActive by mutableStateOf(false)
        private set

    /** Enter clinician mode. Refuses for a user without a clinical role. */
    fun enter(roles: List<String>) {
        if (roles.any { it in CLINICIAN_ROLES }) isActive = true
    }

    fun exit() {
        isActive = false
    }

    /** Drop out when the signed-in user can no longer hold the mode. */
    fun reconcile(roles: List<String>) {
        if (isActive && roles.none { it in CLINICIAN_ROLES }) isActive = false
    }
}

/**
 * Sentinel route for the More grid's "Clinician View" tile. It switches persona
 * instead of navigating, so it is intercepted rather than handed to the
 * NavController — the clinician shell is a different tab bar, not a destination
 * inside the patient one.
 */
const val CLINICIAN_MODE_ROUTE = "__clinician_mode__"

File: IOS/ALAFIA/Core/Auth/ClinicianMode.swift
import Foundation
import Combine

/// Roles that can practise in clinician mode.
///
/// This list lives here rather than inside a view because three places used to
/// need it — the tab bar, the Health hub and the Role screen — and a copy in
/// each is how they drift apart.
enum ClinicianRoles {
    static let all: Set<String> = [
        "physician", "surgeon", "nurse_practitioner",
        "physician_assistant", "resident", "fellow", "attending_physician",
        "cardiologist", "dermatologist", "endocrinologist", "gastroenterologist",
        "neurologist", "oncologist", "pediatrician", "radiologist",
        "general_surgeon", "orthopedic_surgeon", "neurosurgeon",
        "cardiothoracic_surgeon", "plastic_surgeon", "vascular_surgeon",
        "oral_surgeon", "clinical_nurse_specialist", "nurse_anesthetist",
        "nurse_midwife", "charge_nurse", "nurse_administrator",
        "medical_director", "chief_medical_officer",
    ]

    /// True when the user holds any role that can enter clinician mode.
    static func contains(user: User?) -> Bool {
        guard let user else { return false }
        var roles = user.activeRoles ?? []
        if let primary = user.primaryRole { roles.append(primary) }
        return roles.contains { all.contains($0) }
    }

    /// The clinician role the user actually holds, for labelling.
    static func held(by user: User?) -> String? {
        guard let user else { return nil }
        var roles = user.activeRoles ?? []
        if let primary = user.primaryRole { roles.append(primary) }
        return roles.first { all.contains($0) }
    }
}

/// Which persona the app is presenting: the user's own record, or their
/// clinical practice. Switching swaps the entire tab bar rather than adding a
/// screen, so a physician reviewing patients is not navigating past their own
/// meal diary to do it.
@MainActor
final class ClinicianMode: ObservableObject {
    @Published private(set) var isActive = false

    /// Enter clinician mode. Callers pass the current user so a patient account
    /// can never be put into a mode its roles do not allow.
    func enter(as user: User?) {
        guard ClinicianRoles.contains(user: user) else { return }
        isActive = true
    }

    func exit() {
        isActive = false
    }

    /// Drop out of clinician mode when the signed-in user can no longer hold it
    /// — a different account signing in, or a role revoked mid-session.
    func reconcile(with user: User?) {
        if isActive && !ClinicianRoles.contains(user: user) {
            isActive = false
        }
    }
}
